src/Raven.Server/Documents/Expiration/ExpirationStorage.cs

```diff
@@ -246,98 +235,146 @@ private static bool HasPassed(BlittableJsonReaderObject metadata, string metadat
+            foreach (var documentInfo in expired)
             {
-                foreach (var ids in pair.Value)
+                if (documentInfo.Id != null)
                 {
-                    if (ids.Id != null)
+                    try
                     {
-                        try
+                        using (var doc = _database.DocumentsStorage.Get(context, documentInfo.LowerId, DocumentFields.Data, throwOnConflict: true))
                         {
-                            using (var doc = _database.DocumentsStorage.Get(context, ids.LowerId, DocumentFields.Data, throwOnConflict: true))
+                            if (doc != null && doc.TryGetMetadata(out var metadata))
                             {
-                                if (doc != null && doc.TryGetMetadata(out var metadata))
+                                if (HasPassed(metadata, Constants.Documents.Metadata.Expires, currentTime))
                                 {
-                                    if (HasPassed(metadata, Constants.Documents.Metadata.Expires, currentTime))
-                                    {
-                                        _database.DocumentsStorage.Delete(context, ids.LowerId, ids.Id, expectedChangeVector: null);
-                                    }
+                                    _database.DocumentsStorage.Delete(context, documentInfo.LowerId, documentInfo.Id, expectedChangeVector: null);
                                 }
-                                context.Transaction.ForgetAbout(doc);
                             }
-                        }
-                        catch (DocumentConflictException)
-                        {
-                            if (GetConflictedExpiration(context, currentTime, ids.LowerId).AllExpired)
-                                _database.DocumentsStorage.Delete(context, ids.LowerId, ids.Id, expectedChangeVector: null);
-                        }
 
-                        deletionCount++;
+                            context.Transaction.ForgetAbout(doc);
+                        }
                     }
-
-                    expirationTree.MultiDelete(pair.Key, ids.LowerId);
+                    catch (DocumentConflictException)
+                    {
+                        if (GetConflictedExpiration(context, currentTime, documentInfo.LowerId).AllExpired)
+                            _database.DocumentsStorage.Delete(context, documentInfo.LowerId, documentInfo.Id, expectedChangeVector: null);
+                    }
+                    deletionCount++;
                 }
+
+                expirationTree.MultiDelete(documentInfo.Ticks, documentInfo.LowerId);
+                count++;
+
+                if (context.Transaction.InnerTransaction.LowLevelTransaction.TransactionSize > MaxTransactionSize)
+                    break;
             }
 
+            var tx = context.Transaction.InnerTransaction.LowLevelTransaction;
+            tx.OnDispose += _ =>
```

Comment:
This is a new concept. Can you elaborate more on this?

Reply:
This isn't new: https://github.com/ravendb/ravendb/blob/3eab9f8298b16b2559e08657132b1eb10be137a5/src/Raven.Server/Documents/ExecuteRateLimitedOperations.cs#L86